ffs/scheduler/internal/sjstore/sjstore.go

```diff
@@ -360,111 +369,169 @@ func (s *Store) GetStartedDeals(iid ffs.APIID, c cid.Cid) ([]cid.Cid, error) {
+	}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, false, "", fmt.Errorf("querying datastore: %v", err)
+	}
+	var jobs []ffs.StorageJob
+	foundNextPageToken := false
+	if config.NextPageToken == "" {
+		foundNextPageToken = true
+	}
+	done := false
+	more := false
+	nextPageToken := ""
+	for r := range res.Next() {
+		// return an error if there was an error iterating next.
+		if r.Error != nil {
+			return nil, false, "", fmt.Errorf("iter next: %s", r.Error)
+		}
 
-// LatestSuccessfulJobs returns the most recent successful jobs for the specified instance id and cids.
-// If the instance id is ffs.EmptyInstanceID, data for all instances is returned.
-// If no cids are provided, data for all data cids is returned.
-func (s *Store) LatestSuccessfulJobs(iid ffs.APIID, cids ...cid.Cid) []ffs.StorageJob {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+		// if in the last loop we decided we're done, we use this iteration
+		// just to note that there is more data available then break.
+		if done {
+			more = true
+			break
+		}
 
-	return mappedJobs(s.lastSuccessfulJobs, iid, cids...)
-}
+		jobIDString := string(r.Value)
+		jobID := ffs.JobID(jobIDString)
 
-func mappedJobs(m map[ffs.APIID]map[cid.Cid]*ffs.StorageJob, iid ffs.APIID, cids ...cid.Cid) []ffs.StorageJob {
-	var iids []ffs.APIID
-	if iid == ffs.EmptyInstanceID {
-		for iid := range m {
-			iids = append(iids, iid)
+		// if we haven't found the record we need to seek to, continue to the next.
+		if !foundNextPageToken {
+			// additionally, if this is the record we are seeking to, note that we've found it, then continue.
+			if config.NextPageToken == jobIDString {
+				foundNextPageToken = true
+			}
+			continue
 		}
-	} else {
-		iids = append(iids, iid)
-		ensureJobsMap(m, iid)
-	}
 
-	var res []ffs.StorageJob
-	for _, iid := range iids {
-		cidsList := cids
-		if len(cidsList) == 0 {
-			for cid := range m[iid] {
-				cidsList = append(cidsList, cid)
+		// Filter out based on queued/executing etc
+		switch config.Select {
+		case All:
+		case Queued:
+			if _, queued := s.queuedIDs[jobID]; queued {
+				break
+			}
+			continue
+		case Executing:
+			if _, executing := s.executingIDs[jobID]; executing {
+				break
+			}
+			continue
+		case Final:
+			_, queued := s.queuedIDs[jobID]
+			_, executing := s.executingIDs[jobID]
+			if !queued && !executing {
+				break
 			}
+			continue
 		}
 
-		for _, cid := range cidsList {
-			job := m[iid][cid]
-			if job != nil {
-				res = append(res, *job)
-			}
+		job, err := s.get(jobID)
+		if err != nil {
+			return nil, false, "", fmt.Errorf("getting job: %v", err)
+		}
+		jobs = append(jobs, job)
+		nextPageToken = jobIDString
+		if len(jobs) == int(config.Limit) {
+			done = true
 		}
+		continue
 	}
 
-	// Return these newest first.
-	sort.Slice(res, func(a, b int) bool {
-		return res[a].CreatedAt > res[b].CreatedAt
-	})
+	if !more {
+		nextPageToken = ""
+	}
 
-	return res
+	return jobs, more, nextPageToken, nil
 }
```

Comment:
Apart from the `Orders` point, I think this implementation makes sense.
This got quite complex considering how unfortunate making queries with go-datastore is.

Reply:
Yea feels so hard to do something simple. I'd like to start thinking ahead more during the design phase of new code and start with indexes from the beginning in new features.

Reply:
Yeah, totally agree.